File: src/leetcode/p1930/Solution.java
package leetcode.p1930;
// https://leetcode.com/problems/unique-length-3-palindromic-subsequences/

import java.util.*;

/**
 * DP
 * | Time: O(n)
 * | Space: O(n)
 */
public class Solution {
    public int countPalindromicSubsequence(String s) {
        int n= s.length();
        int[] lCnt= new int[26], rCnt= new int[26];
        for (int i=n-1; i >= 1; --i) {
            ++rCnt[s.charAt(i)-'a'];
        }
        Set<String> uniquePalindromes= new HashSet<>();
        for (int i=1; i < n-1; ++i) {
            ++lCnt[s.charAt(i-1)-'a'];
            --rCnt[s.charAt(i)-'a'];
            for (int j=0; j < 26; ++j) {
                if (lCnt[j] > 0 && rCnt[j] > 0) {
                    uniquePalindromes.add("" + (char)('a'+j) + s.charAt(i) + (char)('a'+j));
                }
            }
        }
        return uniquePalindromes.size();
    }
}

File: src/leetcode/cpp/p1930/solution.cpp
// https://leetcode.com/problems/unique-length-3-palindromic-subsequences/
/***
 * Prefix Sum
 * Time: O(26^2*n)
 * Space: O(26n)
 */
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int countPalindromicSubsequence(string s) {
        int n= s.length();
        vector<vector<int>> pSum(26, vector<int>(n,0));
        for (int i=0; i < n; ++i) {
            pSum[s[i]-'a'][i]++;
        }
        for (int i=0; i < 26; ++i) {
            for (int j=1; j < n; ++j) {
                pSum[i][j] += pSum[i][j-1];
            }
        }
        
        int ans= 0;
        for (int a= 0; a < 26; ++a) {
            for (int m=0; m < 26; ++m) {
                for (int i=1; i < n-1; ++i) {
                    if (s[i]-'a' == m) {
                        if (pSum[a][i-1] && pSum[a][n-1]-pSum[a][i] > 0) {
                            ++ans;
                            break;
                        }
                    }
                }
            }
        }
        return ans;
    }
};